solutions/12/script.js

```diff
@@ -0,0 +1,97 @@
+
+// DOM elements
+const song = document.querySelector('.song');
+const artist = document.querySelector('.artist');
+const playBtn = document.querySelector('.play');
+const replayBtn = document.querySelector('.replay');
+const previousBtn = document.querySelector('.previous');
+const nextBtn = document.querySelector('.next');
+const randomBtn = document.querySelector('.random');
+const audioBtn = document.querySelector('.audio');
+const hamburgerBtn = document.querySelector('.hamburger');
+const audioElement = document.querySelector('.audioElement');
+const progress = document.querySelector('.custom-progress')
+
+const musicArrLength = music.length;
+let songIndex = 0;
+progress.value = 0;
+
+// Flags
+let playFlag = false;
+let audioFlag = false;
+let randomFlag = false;
+
+function setSong() {
+    song.textContent = music[songIndex][0];
+    artist.textContent = music[songIndex][1];
+}
+
+function randomSong() {
+    const random = Math.floor(Math.random() * musicArrLength);
+    songIndex = random;
+}
+function nextSong() {
+    randomFlag
+        ? randomSong()
+        : songIndex === musicArrLength - 1
+            ? songIndex = 0
+            : songIndex += 1;
+    setSong();
+}
+
+function previousSong() {
+    randomFlag
+        ? randomSong()
+        : songIndex === 0
+            ? songIndex = musicArrLength - 1
+            : songIndex -= 1;
+    setSong();
+}
+
+function updateProgressBar() {
+    const progressPercentage = audioElement.currentTime / audioElement.duration;
+    progress.value = progressPercentage;
+}
+
+// Event listeners
+previousBtn.addEventListener('click', previousSong);
+nextBtn.addEventListener('click', nextSong);
+randomBtn.addEventListener('click', () => {
+    randomFlag = !randomFlag;
+    randomFlag ? randomBtn.classList.add('active') : randomBtn.classList.remove('active');
+})
+audioBtn.addEventListener('click', () => {
+    if(audioFlag) {
+        audioBtn.classList.remove('fa-volume-up');
+        audioBtn.classList.add('fa-volume-off');
+    } else {
+        audioBtn.classList.remove('fa-volume-off');
+        audioBtn.classList.add('fa-volume-up');
+    }
+    audioFlag = !audioFlag;
+})
+
+playBtn.addEventListener('click', () => {
+    if(playFlag && audioElement.paused) {
+        audioElement.play();
+        playBtn.classList.remove('fa-play');
+        playBtn.classList.add('fa-pause');
+    } else {
+        playBtn.classList.remove('fa-pause');
+        playBtn.classList.add('fa-play');
+        audioElement.pause();
+    }
+    playFlag = !playFlag;
+})
+
+replayBtn.addEventListener('click', () => {
+    audioElement.currentTime = 0;
+    audioElement.play();
+    playBtn.classList.remove('fa-play');
+    playBtn.classList.add('fa-pause');
+})
+
+// Functions to run
+setSong();
+setInterval(updateProgressBar, 10);
```

Comment:
Antipattern you don't know how long will take callback, better is use setTimeout but in the best way for performance is requestAnimationFrame

Reply:
requestAnimationFrame is good option :)